File: Python3/Hard/WrongAnswer/count-subarrays-with-score-less-than-k_6.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        
        # prefix sum
        # sliding window
        # monotonic queue
        # O(n)
        n = len(nums)
        prefix = [0] * (n+1)
        for i in range(n):
            prefix[i+1] = prefix[i] + nums[i]
        # print(prefix)
        res = 0
        q = collections.deque()
        for i in range(n+1):
            while q and prefix[i] <= prefix[q[-1]]:
                q.pop()
            while q and prefix[i] - prefix[q[0]] >= k:
                q.popleft()
            res += len(q)
            q.append(i)
        return res

File: C++/Hard/WrongAnswer/count-subarrays-with-score-less-than-k_6.cpp
class Solution {
public:
    long long countSubarrays(vector<int>& nums, long long k) {
        
        long long res = 0;
        int n = nums.size();
        vector<long long> prefix(n+1, 0);
        for (int i = 1; i <= n; ++i) {
            prefix[i] = prefix[i-1] + nums[i-1];
        }
        deque<int> q;
        for (int i = 0; i <= n; ++i) {
            while (!q.empty() && prefix[i] <= prefix[q.back()]) {
                q.pop_back();
            }
            while (!q.empty() && prefix[i]*q.size() >= prefix[q.front()]*i) {
                q.pop_front();
            }
            q.push_back(i);
            res += q.size()-1;
        }
        return res;
    }
};